### PR TITLE
Add Support For Volunteer & Volunteer Groups in User & Admin Portal

### LINKED ISSUE #2327: GSoC: Add support for Event Volunteer/Group/Membership in User/Admin Portal
**Is your feature request related to a problem? Please describe.**
The Admin Portal & User Portal misses functionality of creating volunteers/ volunteer groups/inviting volunteers, etc.

**Describe the solution you'd like**
In Admin Portal (for Admin/Super Admin/Event Admin) in event Dashboard:
 - Add support for CRUD of Event Volunteers (Individual)
 - Add support for CRUD of Event Volunteer Groups, inviting members, etc.
 - Assign Action Items to Volunteers

In User Portal,
 - User should be able to view upcoming events, and available volunteer groups, join volunteer groups
 - User should be able to accept/reject invitations for joining events & volunteer groups
 - User should be able to view assigned action items.

**Potential internship candidates**

Please read this if you are planning to apply for a Palisadoes Foundation internship 
- https://github.com/PalisadoesFoundation/talawa/issues/359

### PR DESCRIPTION
**Pull request by GlenDsza**
**What kind of change does this PR introduce?**

bugfix, feature, refactoring

**Issue Number:**

Fixes #2327 

**Did you add tests for your changes?**

Yes

**Snapshots/Videos:**

https://github.com/user-attachments/assets/7533459c-95da-44ca-8058-0adcb4a5a3c9

**Summary**

- In Admin, Added Support for Event Volunteer & Groups along with ability to handle Group Membership requests.
- In Admin, integrated action items with volunteer and groups
- In User, Added volunteer dashboard, to volunteer for upcoming events, handle invitations, request to join groups, view assigned action items and view joined volunteer groups.
- In Admin Added volunteer rankings leaderboard, to view most active volunteers and their volunteered hours with various time frame filters.
- All the screens added in Admin & User portal support Sorting Filtering and search by functionalities integrated with backend. 

**Does this PR introduce a breaking change?**

This PR is only supposed to be merged after [API PR](https://github.com/PalisadoesFoundation/talawa-api/pull/2615) gets merged

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-admin/blob/master/CONTRIBUTING.md)?**

Yes 

## Summary by CodeRabbit

- **New Features**
	- Introduced a leaderboard section with new translations for various leaderboard-related texts.
	- Added support for managing volunteer groups and invitations with enhanced user interactions.
	- Implemented new routes for accessing the Fund Campaign Pledge and Volunteer Management components.
	- Enhanced the Volunteer Management interface with new styling and functionality.
	- Added a new component for managing upcoming volunteer events, including search and filter capabilities.
	- Expanded localization support for volunteer-related functionalities across multiple languages.
	- Introduced new interfaces and improved data handling for user-related information.
	- Enhanced action item management with new properties and flexible assignment options.

- **Bug Fixes**
	- Improved error handling and user feedback for various components, including modals and data fetching.

- **Documentation**
	- Updated mock data and test cases to reflect changes in the application structure and functionality.

- **Style**
	- Added new CSS styles to improve the visual presentation of the Volunteer Management and Organization Dashboard components.

- **Tests**
	- Comprehensive test suites added for new components and functionalities to ensure robust coverage of user interactions and error states.
	- Enhanced tests for the `Actions` and `Invitations` components to cover various user interactions and error handling scenarios.

### CODE REVIEW

File: src/screens/OrganizationActionItems/testObject.mocks.ts [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+          creator: {
+            _id: 'creatorId2',
+            firstName: 'John',
+            lastName: 'Doe',
+          },
+        },
+      ],
+    },
+  },
+};
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/screens/OrganizationActionItems/testObject.mocks.ts:402]: test failure is not dependent on fields createdAt

File: src/screens/OrganizationActionItems/testObject.mocks.ts [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+  postCompletionNotes: 'Cmp Notes 1',
+  assignmentDate: new Date('2024-08-27'),
+  dueDate: new Date('2044-08-30'),
+  completionDate: new Date('2044-09-03'),
+  isCompleted: true,
+  event: null,
+  allotedHours: 24,
+  actionItemCategory: actionItemCategory1,
+  ...baseActionItem,
+};
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/screens/OrganizationActionItems/testObject.mocks.ts:182]: these fields dont determine test failure

File: .github/workflows/pull-request.yml [hunk 3]
```diff
@@ -38,7 +38,7 @@ jobs:
       - name: Count number of lines
         run: |
           chmod +x ./.github/workflows/countline.py
-          ./.github/workflows/countline.py --lines 600 --exclude_files src/screens/LoginPage/LoginPage.tsx src/GraphQl/Queries/Queries.ts src/screens/OrgList/OrgList.tsx src/GraphQl/Mutations/mutations.ts src/components/EventListCard/EventListCardModals.tsx src/screens/ManageTag/ManageTag.tsx
+          ./.github/workflows/countline.py --lines 600 --exclude_files src/screens/LoginPage/LoginPage.tsx src/GraphQl/Queries/Queries.ts src/screens/OrgList/OrgList.tsx src/GraphQl/Mutations/mutations.ts src/components/EventListCard/EventListCardModals.tsx src/screens/ManageTag/ManageTag.tsx src/utils/interfaces.ts
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply .github/workflows/pull-request.yml:41]: ideally all interfaces in one file helps

File: src/screens/OrganizationActionItems/OrganizationActionItem.mocks.ts [hunk 4]
```diff
@@ -323,6 +203,7 @@
       variables: {
         actionItemId: 'actionItemId1',
         assigneeId: 'userId1',
+        assigneeType: 'User',
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/screens/OrganizationActionItems/OrganizationActionItem.mocks.ts:206]: done

File: src/GraphQl/Queries/EventVolunteerQueries.ts [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+          image
+        }
+      }
+      group {
+        _id
+        name
+      }
+    }
+  }
+`;
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/GraphQl/Queries/EventVolunteerQueries.ts:118]: not required

File: src/GraphQl/Queries/EventVolunteerQueries.ts [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+      user {
+        _id
+        lastName
+        firstName
+        image
+        email
+      }
+    }
+  }
+`;
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/GraphQl/Queries/EventVolunteerQueries.ts:134]: time frame filter is present

File: src/screens/UserPortal/Volunteer/Actions/Actions.mocks.ts [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+    _id: 'eventId2',
+    title: 'Event 2',
+  },
+  creator: {
+    _id: 'userId1',
+    firstName: 'Wilt',
+    lastName: 'Shepherd',
+  },
+  allottedHours: null,
+};
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/screens/UserPortal/Volunteer/Actions/Actions.mocks.ts:119]: dueDate fields dont determine any test process status

File: src/screens/OrganizationActionItems/ItemModal.test.tsx [hunk 8]
```diff
@@ -295,36 +720,36 @@ describe('Testing ItemModal', () => {
...
-      expect(allotedHours).toHaveValue('0');
+      expect(allottedHours).toHaveValue('0');
     });
 
-    fireEvent.change(allotedHours, { target: { value: '19' } });
+    fireEvent.change(allottedHours, { target: { value: '19' } });
     await waitFor(() => {
-      expect(allotedHours).toHaveValue('19');
+      expect(allottedHours).toHaveValue('19');
     });
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/screens/OrganizationActionItems/ItemModal.test.tsx:753]: Even though suggested is efficient but previous is easy to understand

File: src/screens/OrganizationActionItems/ItemModal.tsx [hunk 9]
```diff
@@ -338,29 +443,132 @@ const ItemModal: FC<InterfaceItemModalProps> = ({
...
+                      /* istanbul ignore next */
+                      handleFormChange('assigneeId', newAssignee?._id ?? '');
+                      setAssigneeUser(newAssignee);
+                    }}
+                    renderInput={(params) => (
+                      <TextField {...params} label={t('assignee')} required />
+                    )}
+                  />
+                </Form.Group>
+              )}
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/screens/OrganizationActionItems/ItemModal.tsx:571]: Autocomplete is an MUI component which doesnt have any such attributes

File: src/utils/interfaces.ts [hunk 10]
```diff
@@ -585,3 +574,111 @@
...
+  }[];
+}
+
+export interface InterfaceCreateVolunteerGroup {
+  name: string;
+  description: string | null;
+  leader: InterfaceUserInfo | null;
+  volunteersRequired: number | null;
+  volunteerUsers: InterfaceUserInfo[];
+}
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/utils/interfaces.ts:627]: done

File: src/utils/interfaces.ts [hunk 11]
```diff
@@ -585,3 +574,111 @@ export interface InterfaceCustomFieldData {
...
+    description: string;
+    volunteers: { _id: string }[];
+  }[];
+  volunteers: {
+    _id: string;
+    user: {
+      _id: string;
+    };
+  }[];
+}
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/utils/interfaces.ts:653]: done

File: src/screens/OrganizationActionItems/testObject.mocks.ts [hunk 12]
```diff
@@ -0,0 +1,402 @@
...
+  postCompletionNotes: 'Cmp Notes 1',
+  assignmentDate: new Date('2024-08-27'),
+  dueDate: new Date('2044-08-30'),
+  completionDate: new Date('2044-09-03'),
+  isCompleted: true,
+  event: null,
+  allottedHours: 24,
+  actionItemCategory: actionItemCategory1,
+  ...baseActionItem,
+};
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/screens/OrganizationActionItems/testObject.mocks.ts:182]: done

File: src/screens/UserPortal/Volunteer/Actions/Actions.tsx [hunk 13]
```diff
@@ -0,0 +1,478 @@
...
+            className={styles.inputField}
+            value={searchValue}
+            onChange={(e) => setSearchValue(e.target.value)}
+            onKeyUp={(e) => {
+              if (e.key === 'Enter') {
+                setSearchTerm(searchValue);
+              } else if (e.key === 'Backspace' && searchValue === '') {
+                setSearchTerm('');
+              }
+            }}
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/screens/UserPortal/Volunteer/Actions/Actions.tsx]: done

**Comment by palisadoes** [comment]
Please make the changes to make CodeRabbit approve the PR

**Comment by palisadoes** [comment]
Please also fix the conflicting file

File: src/screens/UserPortal/Volunteer/Actions/Actions.tsx [hunk 14]
```diff
@@ -0,0 +1,478 @@
...
+  }
+
+  if (actionItemsError) {
+    return (
+      <div className={styles.message} data-testid="errorMsg">
+        <WarningAmberRounded className={styles.icon} fontSize="large" />
+        <h6 className="fw-bold text-danger text-center">
+          {tErrors('errorLoading', { entity: 'Action Items' })}
+          <br />
+          {`${actionItemsError.message}`}
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/screens/UserPortal/Volunteer/Actions/Actions.tsx:153]: done

File: src/screens/UserPortal/Volunteer/Actions/Actions.tsx [hunk 15]
```diff
@@ -0,0 +1,478 @@
...
+          ),
+        }}
+        sx={dataGridStyle}
+        getRowClassName={() => `${styles.rowBackground}`}
+        autoHeight
+        rowHeight={65}
+        rows={actionItems.map((actionItem, index) => ({
+          id: index + 1,
+          ...actionItem,
+        }))}
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/screens/UserPortal/Volunteer/Actions/Actions.tsx:452]: done

File: src/screens/UserPortal/Volunteer/Actions/Actions.tsx [hunk 16]
```diff
@@ -0,0 +1,478 @@
...
+            data-testid="searchBy"
+          />
+          <Button
+            tabIndex={-1}
+            className={`position-absolute z-10 bottom-0 end-0 d-flex justify-content-center align-items-center`}
+            style={{ marginBottom: '10px' }}
+            data-testid="searchBtn"
+          >
+            <Search />
+          </Button>
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/screens/UserPortal/Volunteer/Actions/Actions.tsx:374]: not required

File: src/screens/UserPortal/Volunteer/UpcomingEvents/UpcomingEvents.tsx [hunk 17]
```diff
@@ -0,0 +1,379 @@
...
+                      <h3 data-testid="eventTitle">{title}</h3>
+                      {recurring && (
+                        <Chip
+                          icon={<Circle className={styles.chipIcon} />}
+                          label={'Recurring' as unknown as string}
+                          variant="outlined"
+                          color="primary"
+                          className={`${styles.chip} ${styles.active}`}
+                        />
+                      )}
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/screens/UserPortal/Volunteer/UpcomingEvents/UpcomingEvents.tsx:241]: done

File: src/screens/Leaderboard/Leaderboard.tsx [hunk 18]
```diff
@@ -0,0 +1,372 @@
...
+  },
+};
+
+/**
+ * Component for managing and displaying action items within an organization.
+ *
+ * This component allows users to view, filter, sort, and create action items. It also handles fetching and displaying related data such as action item categories and members.
+ *
+ * @returns The rendered component.
+ */
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/screens/Leaderboard/Leaderboard.tsx:63]: DONE

File: src/screens/OrganizationActionItems/OrganizationActionItems.test.tsx [hunk 19]
```diff
@@ -301,7 +309,8 @@ describe('Testing Organization Action Items Screen', () => {
     expect(searchInput).toBeInTheDocument();
 
     userEvent.type(searchInput, 'John');
-    userEvent.click(screen.getByTestId('searchBtn'));
+    await debounceWait();
+
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/screens/OrganizationActionItems/OrganizationActionItems.test.tsx:313]: done

File: src/screens/UserPortal/Volunteer/Invitations/Invitations.tsx [hunk 20]
```diff
@@ -0,0 +1,301 @@
...
+        </div>
+      </div>
+      {invitations.length < 1 ? (
+        <Stack height="100%" alignItems="center" justifyContent="center">
+          {/* Displayed if no invitations are found */}
+          {t('noInvitations')}
+        </Stack>
+      ) : (
+        invitations.map(
+          (invite: InterfaceVolunteerMembership, index: number) => (
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/screens/UserPortal/Volunteer/Invitations/Invitations.tsx:234]: done

**Comment by SiddheshKukade** [comment]
Taking a look

File: src/screens/UserPortal/Volunteer/Invitations/Invitations.tsx [hunk 21]
```diff
@@ -0,0 +1,301 @@
...
+              className="bg-white p-4  rounded shadow-sm d-flex justify-content-between mb-3"
+              key={index}
+            >
+              <div className="d-flex flex-column gap-2">
+                <div className="fw-bold" data-testid="inviteSubject">
+                  {invite.group ? (
+                    <>Invitation to join volunteer group</>
+                  ) : (
+                    <>Invitation to volunteer for event</>
+                  )}
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/screens/UserPortal/Volunteer/Invitations/Invitations.tsx:245]: done

File: src/screens/UserPortal/Volunteer/Actions/Actions.test.tsx [hunk 22]
```diff
@@ -0,0 +1,221 @@
...
+  ...JSON.parse(JSON.stringify(i18n.getDataByLanguage('en')?.errors ?? {})),
+};
+
+const debounceWait = async (ms = 350): Promise<void> => {
+  await act(() => {
+    return new Promise((resolve) => {
+      setTimeout(resolve, ms);
+    });
+  });
+};
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/screens/UserPortal/Volunteer/Actions/Actions.test.tsx:41]: done

File: src/screens/UserPortal/Volunteer/Actions/Actions.test.tsx [hunk 23]
```diff
@@ -0,0 +1,221 @@
...
+  });
+
+  beforeEach(() => {
+    setItem('userId', 'userId');
+  });
+
+  afterAll(() => {
+    jest.clearAllMocks();
+  });
+
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/screens/UserPortal/Volunteer/Actions/Actions.test.tsx:80]: not required

**Comment by tasneemkoushar** [comment]
@GlenDsza please rebase the branch.

**Comment by tasneemkoushar** [comment]
Please don't merge this PR, I am yet to take synopsis of this PR from Glen. Will take it today and get this one closed

File: src/utils/interfaces.ts [hunk 24]
```diff
@@ -7,6 +7,27 @@ export interface InterfaceUserType {
   };
 }
 
+export interface InterfaceUserInfo {
+  _id: string;
+  firstName: string;
+  lastName: string;
+  image?: string | null;
+}
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply src/utils/interfaces.ts:15]: email is not needed in frontend for most part

File: public/locales/zh/translation.json [hunk 25]
```diff
@@ -733,10 +752,11 @@
...
     "registrants": "注册者",
-    "eventActions": "事件动作",
-    "eventAgendas": "活动议程",
-    "eventStats": "事件统计",
-    "to": "到"
+    "actions": "行动",
+    "agendas": "议程",
+    "statistics": "统计数据",
+    "to": "到",
+    "volunteers": "志愿者"
```
[the remark this answers is not in the document]
Comment (GlenDsza) [reply public/locales/zh/translation.json:759]: done

**Review by tasneemkoushar — APPROVED** [review]

**Pull request merged**